internal/provider/data_source_account.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package provider
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/hashicorp/boundary/api"
+	"github.com/hashicorp/boundary/api/accounts"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+func dataSourceAccount() *schema.Resource {
+	return &schema.Resource{
+		Description: "The boundary_account data source allows you to find a Boundary account.",
+		ReadContext: dataSourceAccountRead,
+
+		Schema: map[string]*schema.Schema{
+			NameKey: {
+				Description:  "The name of the account to retrieve.",
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			AuthMethodIdKey: {
+				Description:  "The auth method ID that will be queried for the account.",
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			IDKey: {
+				Description: "The ID of the retrieved account.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			DescriptionKey: {
+				Description: "The description of the retrieved account.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			TypeKey: {
+				Description: "The type of the account",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			ScopeKey: {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						IDKey: {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						NameKey: {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						TypeKey: {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						DescriptionKey: {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						ParentScopeIdKey: {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceAccountRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	md := meta.(*metaData)
+
+	name := d.Get(NameKey).(string)
+	authMethodId := d.Get(AuthMethodIdKey).(string)
+
+	acl := accounts.NewClient(md.client)
+	accountsList, err := acl.List(ctx, authMethodId,
+		accounts.WithFilter(FilterWithItemNameMatches(name)),
+	)
+	accounts := accountsList.GetItems()
+	if err != nil {
+		return diag.Errorf("error calling list account: %v", err)
+	}
+	if accounts == nil {
+		return diag.Errorf("no accounts found")
+	}
+	if len(accounts) == 0 {
+		return diag.Errorf("no matching account found: %v", err)
+	}
+	if len(accounts) > 1 {
+		return diag.Errorf("error found more than 1 account: %v", err)
```

Comment:
Will `len(accounts) == 0` & `len(accounts) > 1` produce errors from Boundary? I didn't think so. Asking because you always return an error in those cases

Reply:
Chatted with Elim. I was unnecessarily including `err` in these error messages.